### PR TITLE
change variable storage

### PR DESCRIPTION
**Pull request by butterfiolee**
- I switched the variable storage from memory to calldata and vice versa depending on which was cheaper and I added other small changes that save gas.

### CODE REVIEW

File: contracts/exchange/Erc20Escrow.sol [hunk 1]
```diff
@@ -94,6 +94,9 @@ contract Erc20Escrow is IErc20Escrow, EscrowBase {
 
         escrowedByOrder[_orderId].amount = escrowedByOrder[_orderId].amount - _amount;
         IERC20Upgradeable(escrowedByOrder[_orderId].token).transfer(_receiver, _amount);
+        if (escrowedByOrder[_orderId].amount == 0) {
```
Comment (gcbsumid) [inline contracts/exchange/Erc20Escrow.sol:97]: I don't like this because you're only doing the deleting the token address for withdraw, but not withdrawBatch().

Realistically, I think that to make it way cheaper and remove "token" redundancy, it's better to change escrowedByOrder from mapping(orderId => EscrowAmount) to:
**mapping(tokenAddress => mapping(orderId => amount))** 

this way, instead of storing a whole struct {token, amount}, it uses token as an index and it only stores {amount}. 

You'll have to update all of erc20Escrow.sol though for every escrowByOrder though

File: contracts/exchange/NftEscrow.sol [hunk 2]
```diff
@@ -85,7 +85,6 @@ contract NftEscrow is INftEscrow, EscrowBase, ERC1155HolderUpgradeable, ERC721Ho
         // We don't need to store how much was escrowed because we keep track of the order data in 
         // the orderbook.
         if (escrowedAmounts[_orderId] == 0) {
-            delete escrowedAmounts[_orderId];
```
Comment (gcbsumid) [inline contracts/exchange/NftEscrow.sol:88]: why did you remove this? Is it cheaper without this?

File: contracts/exchange/Orderbook.sol [hunk 3]
```diff
@@ -156,11 +154,11 @@ contract Orderbook is IOrderbook, ManagerBase {
                 amountSpentOnOrder = orders[_orderIds[i]].price * orderAmounts[i];
                 
                 // Check if the transaction is still under the Max Spend
-                if (maxSpend == 0) {
+                if (maxSpend >= amountSpentOnOrder) {
```
Comment (gcbsumid) [inline contracts/exchange/Orderbook.sol:157]: When you moved it, was it cheaper? I assume maxSpend => amountSpentOnOrder is the most common path which is why you reordered it, is that correct?

File: contracts/exchange/RoyaltyManager.sol [hunk 4]
```diff
@@ -139,8 +139,8 @@ contract RoyaltyManager is IRoyaltyManager, ManagerBase {
     }
 
     function buyOrderRoyalties(
-        LibOrder.AssetData calldata _asset,
-        uint256[] memory amountPerOrder
+        LibOrder.AssetData memory _asset,
```
Comment (gcbsumid) [inline contracts/exchange/RoyaltyManager.sol:142]: why did you flip this instead of just making both 'calldata'? I'm curious

**Review by gcbsumid — CHANGES_REQUESTED** [review]

File: contracts/exchange/NftEscrow.sol (hunk 2 above)
Comment (butterfiolee) [reply contracts/exchange/NftEscrow.sol:88]: Yes, because all deleting escrowedAmounts[_orderId] does is set the uint256 value to zero, and because of the if statement, it's already zero, so you don't get a refund for it (or you've already been refunded and you don't get another refund for it). I ran a gas test to verify it.

File: contracts/exchange/Orderbook.sol (hunk 3 above)
Comment (butterfiolee) [reply contracts/exchange/Orderbook.sol:157]: Correct.

File: contracts/exchange/RoyaltyManager.sol (hunk 4 above)
Comment (butterfiolee) [reply contracts/exchange/RoyaltyManager.sol:142]: Surprisingly, switching the storage of the AssetData object to memory was cheaper for both of these functions.

File: contracts/exchange/Erc20Escrow.sol (hunk 1 above)
Comment (gcbsumid) [reply contracts/exchange/Erc20Escrow.sol:97]: Ah, nevermind with this. I realized why I created the structure in the first place.

Comment (gcbsumid) [reply contracts/exchange/Erc20Escrow.sol:97]: If you're going to delete the 'token', you might as well just delete the whole struct escrowedByOrder[orderId] instead of just deleting the token.

Make sure you delete it in withdrawBatch() as well. And check by how much this actually saves.

File: contracts/exchange/NftEscrow.sol (hunk 2 above)
Comment (gcbsumid) [reply contracts/exchange/NftEscrow.sol:88]: Sounds good.

File: contracts/exchange/Orderbook.sol (hunk 3 above)
Comment (gcbsumid) [reply contracts/exchange/Orderbook.sol:157]: sounds good.

File: contracts/exchange/RoyaltyManager.sol (hunk 4 above)
Comment (gcbsumid) [reply contracts/exchange/RoyaltyManager.sol:142]: Interesting, maybe it's cheaper to use "memory" for structures? Can you check other places where we use 'calldata' for structures like LibOrder.AssetData and see if changing it to memory makes that function call cheaper?

File: contracts/exchange/Erc20Escrow.sol (hunk 1 above)
Comment (butterfiolee) [reply contracts/exchange/Erc20Escrow.sol:97]: This is the same deal as in NftEscrow, deleting the whole struct just means setting  .token to the zero address and .amount to zero (which is already zero), so it saves more gas to just delete the token specifically.

withdrawBatch is tricky because it runs the loop of updating the mapping and deleting the variables before sending a the transfer function at the end. And you can't transfer the token if you've already deleted the the tokenAddress. And according to the tests, declaring a local address variable costs quite a bit more than the amount refunded by deleting the tokenAddresses- UNLESS you have a long list of orderIds. So most of the time, it's not worth it.

File: contracts/exchange/RoyaltyManager.sol (hunk 4 above)
Comment (butterfiolee) [reply contracts/exchange/RoyaltyManager.sol:142]: Yeah I checked it for all of them, it was sometimes cheaper and sometimes more expensive for structures to use memory. All of them already use whichever one's cheaper.

File: contracts/exchange/Erc20Escrow.sol (hunk 1 above)
Comment (gcbsumid) [reply contracts/exchange/Erc20Escrow.sol:97]: If that's the case, then you shouldn't delete just in the single Withdraw function. The gas savings is not worth the inconsistency between withdraw and withdrawBatch. The maintainability will go down because people will continually ask in the future why something is done in withdraw, but not withdraw batch. The inconsistency may also create future bugs because something gets deleted in withdraw but not withdrawBatch and it won't be common knowledge.

File: contracts/exchange/RoyaltyManager.sol (hunk 4 above)
Comment (gcbsumid) [reply contracts/exchange/RoyaltyManager.sol:142]: If that's the case, we should just be consistent and use calldata for all function parameters instead of memory even if memory is slightly cheaper. It shouldn't be too much more expensive, and I think the benefits to consistency is much better than a very slight gas savings.

**Review by gcbsumid — APPROVED** [review]
